contrib/db_pools/lib/src/pool.rs

```diff
@@ -0,0 +1,235 @@
+use rocket::async_trait;
+
+/// This trait is implemented on connection pool types that can be used with the
+/// [`Database`] derive macro.
+///
+/// `Pool` determines how the connection pool is initialized from configuration,
+/// such as a connection string and optional pool size, along with the returned
+/// `Connection` type.
+///
+/// Implementations of this trait should use `async_trait`.
+///
+/// ## Example
+///
+/// ```
+/// #[derive(Debug)]
+/// struct Error { /* ... */ }
+/// # impl std::fmt::Display for Error {
+/// #     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// #         unimplemented!("example")
+/// #     }
+/// # }
+/// # impl std::error::Error for Error { }
+///
+/// struct Pool { /* ... */ }
+/// struct Connection { /* .. */ }
+///
+/// #[rocket::async_trait]
+/// impl rocket_db_pools::Pool for Pool {
+///     type Connection = Connection;
+///     type Config = rocket_db_pools::Config;
+///     type InitError = Error;
+///     type GetError = Error;
+///
+///     async fn initialize(config: Self::Config) -> Result<Self, Self::InitError> {
+///         unimplemented!("example")
+///     }
+///
+///     async fn get(&self) -> Result<Connection, Self::GetError> {
+///         unimplemented!("example")
+///     }
+/// }
+/// ```
+///
+/// [`Database`]: crate::Database
+#[async_trait]
+pub trait Pool: Sized + Send + Sync + 'static {
+    /// The type returned by get().
+    type Connection;
+
+    /// The configuration type this database pool expects.
+    type Config: rocket::serde::DeserializeOwned + Send;
+
+    /// The error type returned by `initialize`.
+    type InitError: std::error::Error;
+
+    /// The error type returned by `get`.
+    type GetError: std::error::Error;
+
+    /// Constructs a pool from a [Value](rocket::figment::value::Value).
+    ///
+    /// It is up to each implementor of `Pool` to define its accepted
+    /// configuration value(s) via the `Config` associated type.  Most
+    /// integrations provided in `rocket_db_pools` use [`Config`], which
+    /// accepts a (required) `url` and an (optional) `pool_size`.
+    ///
+    /// ## Errors
+    ///
+    /// This method returns an error if the configuration is not compatible, or
+    /// if creating a pool failed due to an unavailable database server,
+    /// insufficient resources, or another database-specific error.
+    ///
+    /// [`Config`]: crate::Config
+    async fn initialize(config: Self::Config) -> Result<Self, Self::InitError>;
+
+    /// Asynchronously gets a connection from the factory or pool.
+    ///
+    /// ## Errors
+    ///
+    /// This method returns an error if a connection could not be retrieved,
+    /// such as a preconfigured timeout elapsing or when the database server is
+    /// unavailable.
+    async fn get(&self) -> Result<Self::Connection, Self::GetError>;
+}
+
+#[cfg(feature = "deadpool-postgres")]
+#[async_trait]
+impl crate::Pool for deadpool_postgres::Pool {
```

Comment:
```suggestion
impl Pool for deadpool_postgres::Pool {
```

Here and everywhere else.